src/vterm.c

```diff
@@ -0,0 +1,474 @@
+    buffer[2] = 0x80 | (codepoint & 0x3F);
+    return 3;
+  }
+
+  if (codepoint >= 0x10000 && codepoint <= 0x10FFFF) {
+    buffer[0] = 0xF0 | (codepoint >> 18);
+    buffer[1] = 0x80 | ((codepoint >> 12) & 0x3F);
+    buffer[2] = 0x80 | ((codepoint >> 6) & 0x3F);
+    buffer[3] = 0x80 | (codepoint & 0x3F);
+    return 4;
+  }
+  return 0;
+}
+
+static bool
+utf8_to_codepoint(const unsigned char buffer[4], const size_t len,
+                       uint32_t *codepoint) {
+  *codepoint = 0;
+  if (len == 1 && buffer[0] <= 0x7F) {
+    *codepoint = buffer[0];
+    return true;
+  }
+  if (len == 2 && (buffer[0] >= 0xC0 && buffer[0] <= 0xDF) &&
+      (buffer[1] >= 0x80 && buffer[1] <= 0xBF)) {
+    *codepoint = buffer[0] & 0x1F;
+    *codepoint = *codepoint << 6;
+    *codepoint = *codepoint | (buffer[1] & 0x3F);
+    return true;
+  }
+  if (len == 3 && (buffer[0] >= 0xE0 && buffer[0] <= 0xEF) &&
+      (buffer[1] >= 0x80 && buffer[1] <= 0xBF) &&
+      (buffer[2] >= 0x80 && buffer[2] <= 0xBF)) {
+    *codepoint = buffer[0] & 0xF;
+    *codepoint = *codepoint << 6;
+    *codepoint = *codepoint | (buffer[1] & 0x3F);
+    *codepoint = *codepoint << 6;
+    *codepoint = *codepoint | (buffer[2] & 0x3F);
+    return true;
+  }
+  if (len == 4 && (buffer[0] >= 0xF0 && buffer[0] <= 0xF7) &&
+      (buffer[1] >= 0x80 && buffer[1] <= 0xBF) &&
+      (buffer[2] >= 0x80 && buffer[2] <= 0xBF) &&
+      (buffer[3] >= 0x80 && buffer[3] <= 0xBF)) {
+    *codepoint = buffer[0] & 7;
+    *codepoint = *codepoint << 6;
+    *codepoint = *codepoint | (buffer[1] & 0x3F);
+    *codepoint = *codepoint << 6;
+    *codepoint = *codepoint | (buffer[2] & 0x3F);
+    *codepoint = *codepoint << 6;
+    *codepoint = *codepoint | (buffer[3] & 0x3F);
+    return true;
+  }
+
+  return false;
+}
+
+int row_to_linenr(Term *term, int row) {
+  return row != INT_MAX ? row + (int)term->sb_current + 1 : INT_MAX;
+}
+
+static void
+fetch_cell(Term *term, int row, int col, VTermScreenCell *cell) {
+  if (row < 0) {
+    VtermScrollbackLine *sbrow = term->sb_buffer[-row - 1];
+    if ((size_t)col < sbrow->cols) {
+      *cell = sbrow->cells[col];
+    } else {
+      // fill the pointer with an empty cell
+      VTermColor fg, bg;
+      VTermState *state = vterm_obtain_state(term->vt);
+      vterm_state_get_default_colors(state, &fg, &bg);
+
+      *cell = (VTermScreenCell){.chars = {0}, .width = 1, .bg = bg};
+    }
+  } else {
+    vterm_screen_get_cell(term->vts, (VTermPos){.row = row, .col = col}, cell);
+  }
+}
+
+size_t get_col_offset(Term *term, int row, int end_col) {
+  int col = 0;
+  size_t offset = 0;
+  unsigned char buf[4];
+
+  while (col < end_col) {
+    VTermScreenCell cell;
+    fetch_cell(term, row, col, &cell);
+    if (cell.chars[0]) {
+      if (cell.width > 1) {
+        offset += cell.width - 1;
+      }
+    }
+    col += cell.width;
+  }
+  return offset;
+}
+
+static bool compare_cells(VTermScreenCell *a, VTermScreenCell *b) {
```

Comment:
This too is almost identical to an internal `libvterm` function - might be best to expose and use that.